File: yumi/sdk/csharp/ToolArguments.cs
using System.Text.Json;

namespace Yumi;

/// <summary>
/// Type-safe accessor for tool call arguments backed by a <see cref="JsonElement"/>.
/// </summary>
public sealed class ToolArguments
{
    private readonly JsonElement _raw;

    public ToolArguments(JsonElement raw)
    {
        _raw = raw;
    }

    public JsonElement Raw => _raw;

    public string GetString(string key, string fallback = "")
    {
        if (_raw.TryGetProperty(key, out var el) && el.ValueKind == JsonValueKind.String)
            return el.GetString() ?? fallback;
        return fallback;
    }

    public int GetInt(string key, int fallback = 0)
    {
        if (_raw.TryGetProperty(key, out var el) && el.TryGetInt32(out var v))
            return v;
        return fallback;
    }

    public double GetDouble(string key, double fallback = 0.0)
    {
        if (_raw.TryGetProperty(key, out var el) && el.TryGetDouble(out var v))
            return v;
        return fallback;
    }

    public bool GetBool(string key, bool fallback = false)
    {
        if (_raw.TryGetProperty(key, out var el))
        {
            if (el.ValueKind == JsonValueKind.True) return true;
            if (el.ValueKind == JsonValueKind.False) return false;
        }
        return fallback;
    }
}

File: yumi/sdk/java/src/main/java/io/yumi/ToolArguments.java
package io.yumi;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

/**
 * Type-safe accessor for tool call arguments backed by a Gson JsonObject.
 */
public class ToolArguments {
    private final JsonObject raw;

    public ToolArguments(JsonObject raw) {
        this.raw = raw != null ? raw : new JsonObject();
    }

    public JsonObject getRaw() { return raw; }

    public String getString(String key, String fallback) {
        JsonElement e = raw.get(key);
        if (e == null || !e.isJsonPrimitive()) return fallback;
        return e.getAsString();
    }

    public String getString(String key) {
        return getString(key, "");
    }

    public int getInt(String key, int fallback) {
        JsonElement e = raw.get(key);
        if (e == null || !e.isJsonPrimitive()) return fallback;
        try { return e.getAsInt(); } catch (Exception ex) { return fallback; }
    }

    public double getDouble(String key, double fallback) {
        JsonElement e = raw.get(key);
        if (e == null || !e.isJsonPrimitive()) return fallback;
        try { return e.getAsDouble(); } catch (Exception ex) { return fallback; }
    }

    public boolean getBoolean(String key, boolean fallback) {
        JsonElement e = raw.get(key);
        if (e == null || !e.isJsonPrimitive()) return fallback;
        try { return e.getAsBoolean(); } catch (Exception ex) { return fallback; }
    }
}
